lgc/patch/NggPrimShader.cpp

```diff
@@ -2745,89 +2650,110 @@ Function *NggPrimShader::mutateEsToVariant(Module *module, StringRef entryName,
-            expValues[0] = m_builder->CreateBitCast(expValues[0], m_builder->getFloatTy());
-            expValues[1] = m_builder->CreateBitCast(expValues[1], m_builder->getFloatTy());
-            expValues[2] = m_builder->CreateBitCast(expValues[2], m_builder->getFloatTy());
-            expValues[3] = m_builder->CreateBitCast(expValues[3], m_builder->getFloatTy());
-          }
+  //
+  // Create ES-partial to do deferred vertex export after NGG culling
+  //
 
-          Value *expValue = UndefValue::get(FixedVectorType::get(Type::getFloatTy(*m_context), 4));
-          for (unsigned i = 0; i < 4; ++i)
-            expValue = m_builder->CreateInsertElement(expValue, expValues[i], i);
+  // NOTE: Here, we just mutate original ES to do deferred vertex export. We add vertex position data as an additional
+  // input (the first argument). This could avoid re-fetching it since we already get the data before NGG culling.
+  auto esDeferredVertexExportFunc = addFunctionArgs(esEntryPoint, nullptr, positionTy);
+  esDeferredVertexExportFunc->setName(lgcName::NggEsDeferredVertexExport);
 
-          if (expPos) {
-            // Last position export that needs "done" flag
-            lastExport = expDataSet.size();
-          }
+  position = esDeferredVertexExportFunc->getArg(0); // The first argument is vertex position data
+  assert(position->getType() == positionTy);
 
-          ExpData expData = {expTarget, channelMask, false, expValue};
-          expDataSet.push_back(expData);
-        }
+  for (auto func : expFuncs) {
+    for (auto user : func->users()) {
+      CallInst *const call = dyn_cast<CallInst>(user);
+      assert(call);
+
+      if (call->getParent()->getParent() != esDeferredVertexExportFunc)
+        continue; // Export call doesn't belong to targeted function, skip
 
-        expCalls.push_back(call);
+      unsigned exportTarget = cast<ConstantInt>(call->getArgOperand(0))->getZExtValue();
+      if (exportTarget == EXP_TARGET_POS_0) {
+        // Replace vertex position data
+        m_builder->SetInsertPoint(call);
+        call->setArgOperand(2, m_builder->CreateExtractElement(position, static_cast<uint64_t>(0)));
+        call->setArgOperand(3, m_builder->CreateExtractElement(position, 1));
+        call->setArgOperand(4, m_builder->CreateExtractElement(position, 2));
+        call->setArgOperand(5, m_builder->CreateExtractElement(position, 3));
       }
     }
   }
-  assert(expDataSet.size() == expCount);
-
-  // Set "done" flag for last position export
-  if (lastExport != InvalidValue)
-    expDataSet[lastExport].doneFlag = true;
-
-  // Construct exported data
-  unsigned i = 0;
-  for (auto &expDataElement : expDataSet) {
-    expData = m_builder->CreateInsertValue(expData, expDataElement.expValue, i++);
-    expDataElement.expValue = nullptr;
-  }
 
-  // Insert new "return" instruction
-  m_builder->CreateRet(expData);
+  // Original ES is no longer needed
+  esEntryPoint->dropAllReferences();
```

Comment:
Don't use `dropAllReferences`, here or elsewhere in this function. There shouldn't be any uses of this left anyway, but if there are, then `dropAllReferences` will leave those uses in an invalid state. Better to just place an `assert(esEntryPoint->use_empty());` here.

Reply:
Done. Thank you.